jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DelayedHandler.java

```diff
@@ -163,75 +197,146 @@ protected Callback getCallback()
+
+                // The handle was rejected, so write the error using the original potentially unwrapped request/response/callback
+                Response.writeError(getRequest(), getResponse(), getCallback(), HttpStatus.NOT_FOUND_404);
             }
             catch (Throwable t)
             {
+                // The handle failed, so write the error using the original potentially unwrapped request/response/callback
                 Response.writeError(getRequest(), getResponse(), getCallback(), t);
             }
+            // return false to indicate the passed request/response/callback were not used.
+            return false;
         }
 
         protected abstract void delay() throws Exception;
     }
 
-    protected static class UntilContentDelayedProcess extends DelayedProcess
+    /**
+     * Delay dispatch until all content or an effective buffer size is reached
+     */
+    protected static class UntilContentDelayedProcess extends DelayedProcess implements Runnable
     {
-        public UntilContentDelayedProcess(Handler handler, Request request, Response response, Callback callback)
+        private final Deque<Content.Chunk> _chunks = new ArrayDeque<>();
+        private final int _maxSize;
+        private int _estimatedSize;
+
+        /**
+         * @param handler The next handler
+         * @param request The delayed request
+         * @param response The delayed response
+         * @param callback The delayed callback
+         * @param maxSize The maximum size to buffer before dispatching to the next handler;
+         *                or -1 to use {@link HttpConnectionFactory#getInputBufferSize()}
+         */
+        public UntilContentDelayedProcess(Handler handler, Request request, Response response, Callback callback, int maxSize)
         {
             super(handler, request, response, callback);
+            _maxSize = maxSize < 0 ? request.getConnectionMetaData().getConnector().getConnectionFactory(HttpConnectionFactory.class).getInputBufferSize() : maxSize;
         }
 
         @Override
         protected void delay()
         {
-            Content.Chunk chunk = super.getRequest().read();
-            if (chunk == null)
-            {
-                getRequest().demand(org.eclipse.jetty.util.thread.Invocable.from(InvocationType.NON_BLOCKING, this::onContent));
-            }
-            else
+            read(false);
+        }
+
+        protected void onContentAvailable()
+        {
+            read(true);
+        }
+
+        protected void read(boolean execute)
+        {
+            while (true)
             {
-                RewindChunkRequest request = new RewindChunkRequest(getRequest(), chunk);
-                try
+                Content.Chunk chunk = super.getRequest().read();
+                if (chunk == null)
+                {
+                    getRequest().demand(org.eclipse.jetty.util.thread.Invocable.from(InvocationType.NON_BLOCKING, this::onContentAvailable));
```

Comment:
There may be more invocations of this line than the only one time `run()` is called, so I would switch the 2: here call `demand(this)`, and when it's time to call the next `Handler` use a lambda.

Can't this code be written with a `Promise` like form and multipart, and use the same `AtomicInteger(2)` technique instead of calling `execute()`?

Reply:
OK for the inversion of runnable.   However the atomicInteger style doesn't  really work here as read is a pull API and we don't need to check for a callback to know we are all done withing the handler thread.